File: helping-hand/backend/modules/auth/controller.js
const bcrypt = require("bcryptjs");
const jwt = require("jsonwebtoken");

const User = require("./model");

const JWT_SECRET = process.env.JWT_SECRET;
const JWT_EXPIRES_IN = process.env.JWT_EXPIRES_IN || "7d";

function signToken(user) {
  if (!JWT_SECRET) throw new Error("JWT_SECRET is not set");

  return jwt.sign({ id: String(user._id), role: user.role }, JWT_SECRET, {
    expiresIn: JWT_EXPIRES_IN,
  });
}

function safeUser(user) {
  return {
    id: user._id,
    name: user.name,
    email: user.email,
    phone: user.phone,
    role: user.role,
    profilePicture: user.profilePicture,
    isVerified: user.isVerified,
    isBlocked: user.isBlocked,
    bio: user.bio,
    address: user.address,
    city: user.city,
    location: user.location,
    ratingAvg: user.ratingAvg,
    totalDonations: user.totalDonations,
    totalReceived: user.totalReceived,
    createdAt: user.createdAt,
    updatedAt: user.updatedAt,
  };
}

exports.register = async (req, res) => {
  try {
    const { name, email, phone, password, role } = req.body;

    if (!name || !email || !password) {
      return res
        .status(400)
        .json({ message: "name, email and password are required" });
    }

    const existing = await User.findOne({ email: email.toLowerCase().trim() });
    if (existing) return res.status(409).json({ message: "Email already exists" });

    const hashedPassword = await bcrypt.hash(password, 10);

    const user = await User.create({
      name: name.trim(),
      email: email.toLowerCase().trim(),
      phone,
      password: hashedPassword,
      role,
    });

    const token = signToken(user);

    return res.status(201).json({ token, user: safeUser(user) });
  } catch (error) {
    return res.status(500).json({ message: error.message });
  }
};

exports.login = async (req, res) => {
  try {
    const { email, password } = req.body;

    if (!email || !password) {
      return res
        .status(400)
        .json({ message: "email and password are required" });
    }

    const user = await User.findOne({ email: email.toLowerCase().trim() });
    if (!user) return res.status(401).json({ message: "Invalid credentials" });
    if (user.isBlocked) return res.status(403).json({ message: "Account is blocked" });

    const ok = await bcrypt.compare(password, user.password);
    if (!ok) return res.status(401).json({ message: "Invalid credentials" });

    const token = signToken(user);

    return res.json({ token, user: safeUser(user) });
  } catch (error) {
    return res.status(500).json({ message: error.message });
  }
};

exports.forgotPassword = async (req, res) => {
  try {
    const { email } = req.body;

    if (!email) return res.status(400).json({ message: "email is required" });

    const user = await User.findOne({ email: email.toLowerCase().trim() });
    if (!user) return res.status(404).json({ message: "User not found" });

    const otp = Math.floor(100000 + Math.random() * 900000);
    const otpExpiry = new Date(Date.now() + 10 * 60 * 1000);

    user.otp = String(otp);
    user.otpExpiry = otpExpiry;

    await user.save();

    // Testing: return OTP in response
    return res.json({ message: "OTP generated", otp, otpExpiry });
  } catch (error) {
    return res.status(500).json({ message: error.message });
  }
};

exports.verifyOtp = async (req, res) => {
  try {
    const { email, otp } = req.body;

    if (!email || !otp) {
      return res.status(400).json({ message: "email and otp are required" });
    }

    const user = await User.findOne({ email: email.toLowerCase().trim() });
    if (!user) return res.status(404).json({ message: "User not found" });

    const now = Date.now();
    if (!user.otp || !user.otpExpiry) {
      return res.status(400).json({ message: "OTP not set" });
    }

    const isOtpMatch = String(user.otp) === String(otp);
    const isNotExpired = user.otpExpiry.getTime() > now;

    if (!isOtpMatch) return res.status(400).json({ message: "Invalid OTP" });
    if (!isNotExpired) return res.status(400).json({ message: "OTP expired" });

    return res.json({ message: "OTP verified" });
  } catch (error) {
    return res.status(500).json({ message: error.message });
  }
};

exports.resetPassword = async (req, res) => {
  try {
    const { email, otp, newPassword } = req.body;

    if (!email || !otp || !newPassword) {
      return res.status(400).json({ message: "email, otp and newPassword are required" });
    }

    const user = await User.findOne({ email: email.toLowerCase().trim() });
    if (!user) return res.status(404).json({ message: "User not found" });

    if (!user.otp || !user.otpExpiry) {
      return res.status(400).json({ message: "OTP not set" });
    }

    const isOtpMatch = String(user.otp) === String(otp);
    const isNotExpired = user.otpExpiry.getTime() > Date.now();

    if (!isOtpMatch) return res.status(400).json({ message: "Invalid OTP" });
    if (!isNotExpired) return res.status(400).json({ message: "OTP expired" });

    const hashedPassword = await bcrypt.hash(newPassword, 10);
    user.password = hashedPassword;

    // clear otp fields
    user.otp = undefined;
    user.otpExpiry = undefined;

    await user.save();

    return res.json({ message: "Password reset successful" });
  } catch (error) {
    return res.status(500).json({ message: error.message });
  }
};

File: helping-hand/backend/modules/auth/controller.py
import bcrypt
import jwt
from datetime import datetime, timedelta, timezone
import random

from bson import ObjectId

from config.settings import Settings
from modules.auth.model import (
    create_user,
    find_user_by_email,
    find_user_by_id,
    update_user,
)


def _utcnow():
    return datetime.now(timezone.utc)


def _generate_jwt(user_id: str, role: str) -> str:
    if not Settings.JWT_SECRET:
        raise RuntimeError("JWT_SECRET is not set")

    exp = _utcnow() + timedelta(minutes=Settings.JWT_EXP_MINUTES)
    payload = {"user_id": user_id, "role": role, "exp": exp}
    token = jwt.encode(payload, Settings.JWT_SECRET, algorithm=Settings.JWT_ALGORITHM)
    # pyjwt may return str already; keep normalized
    return token if isinstance(token, str) else token.decode("utf-8")


def register():
    from flask import request, jsonify

    data = request.get_json(force=True) or {}

    required = ["name", "email", "password", "phone", "role"]
    missing = [k for k in required if not data.get(k)]
    if missing:
        return jsonify({"error": f"Missing fields: {', '.join(missing)}"}), 400

    if data.get("role") not in ["donor", "recipient", "NGO", "admin"]:
        return jsonify({"error": "Invalid role"}), 400

    if find_user_by_email(data["email"]):
        return jsonify({"error": "Email already registered"}), 409

    hashed = bcrypt.hashpw(data["password"].encode("utf-8"), bcrypt.gensalt())

    user_doc = {
        "name": data["name"],
        "email": data["email"],
        "password": hashed.decode("utf-8"),
        "phone": data["phone"],
        "role": data["role"],
        "profile_picture": data.get("profile_picture", ""),
        "is_verified": data.get("is_verified", False),
        "is_blocked": data.get("is_blocked", False),
        "otp": "",
        "otp_expiry": None,
        "bio": data.get("bio", ""),
        "address": data.get("address", ""),
        "city": data.get("city", ""),
        "location": data.get("location", {"lat": 0.0, "lng": 0.0}),
        "rating_avg": data.get("rating_avg", 0.0),
        "total_donations": data.get("total_donations", 0),
        "total_received": data.get("total_received", 0),
        "created_at": _utcnow(),
    }

    inserted_id = create_user(user_doc)
    token = _generate_jwt(str(inserted_id), user_doc["role"])

    return jsonify({"token": token, "user_id": str(inserted_id), "role": user_doc["role"]}), 201


def login():
    from flask import request, jsonify

    data = request.get_json(force=True) or {}
    email = data.get("email")
    password = data.get("password")

    if not email or not password:
        return jsonify({"error": "email and password are required"}), 400

    user = find_user_by_email(email)
    if not user:
        return jsonify({"error": "Invalid credentials"}), 401

    if user.get("is_blocked"):
        return jsonify({"error": "Account is blocked"}), 403

    stored_hash = user.get("password")
    if not stored_hash:
        return jsonify({"error": "Invalid credentials"}), 401

    if not bcrypt.checkpw(password.encode("utf-8"), stored_hash.encode("utf-8")):
        return jsonify({"error": "Invalid credentials"}), 401

    token = _generate_jwt(str(user["_id"]), user.get("role"))
    return jsonify({"token": token, "user_id": str(user["_id"]), "role": user.get("role")}), 200


def forgot_password():
    from flask import request, jsonify

    data = request.get_json(force=True) or {}
    email = data.get("email")
    if not email:
        return jsonify({"error": "email is required"}), 400

    user = find_user_by_email(email)
    if not user:
        # Avoid user enumeration
        return jsonify({"message": "If that email exists, an OTP will be sent."}), 200

    otp = f"{random.randint(0, 999999):06d}"
    otp_expiry = _utcnow() + timedelta(minutes=10)

    update_user(str(user["_id"]), {"otp": otp, "otp_expiry": otp_expiry})

    # In production, send OTP via email/SMS. Here we return for testing.
    return jsonify({"message": "OTP generated", "otp": otp, "otp_expiry": otp_expiry.isoformat()}), 200


def verify_otp():
    from flask import request, jsonify

    data = request.get_json(force=True) or {}
    email = data.get("email")
    otp = data.get("otp")

    if not email or not otp:
        return jsonify({"error": "email and otp are required"}), 400

    user = find_user_by_email(email)
    if not user:
        return jsonify({"error": "Invalid OTP"}), 400

    if user.get("otp") != str(otp):
        return jsonify({"error": "Invalid OTP"}), 400

    expiry = user.get("otp_expiry")
    if not expiry or expiry <= _utcnow():
        return jsonify({"error": "OTP expired"}), 400

    return jsonify({"message": "OTP verified"}), 200


def reset_password():
    from flask import request, jsonify

    data = request.get_json(force=True) or {}
    email = data.get("email")
    otp = data.get("otp")
    new_password = data.get("new_password")

    if not email or not otp or not new_password:
        return jsonify({"error": "email, otp and new_password are required"}), 400

    user = find_user_by_email(email)
    if not user:
        return jsonify({"error": "Invalid OTP"}), 400

    if user.get("otp") != str(otp):
        return jsonify({"error": "Invalid OTP"}), 400

    expiry = user.get("otp_expiry")
    if not expiry or expiry <= _utcnow():
        return jsonify({"error": "OTP expired"}), 400

    hashed = bcrypt.hashpw(new_password.encode("utf-8"), bcrypt.gensalt())

    update_user(
        str(user["_id"]),
        {
            "password": hashed.decode("utf-8"),
            "otp": "",
            "otp_expiry": None,
        },
    )

    return jsonify({"message": "Password updated"}), 200

